### DISCUSSION TITLE
tests fail after HPC build

### DISCUSSION THREAD

**Discussion by heinono1** [idaholab/moose#21218] [Q&A Getting Started] [answered]
I am building MOOSE on Ubuntu20.04 with newly installed c, g++, gfortran compilers and openmpi wrappers. Because of restrictions at work, I cannot access the inl.gov domain for downloads, but pulled the dist from github and followed the HPC install instructions. After PETSc and libmesh builds I built moose but the test goes wrong: ModuleNotFoundError: No module named "hit". There is also an undefined symbol (_ZSt28__throw_bad_array_new_lengthv). Not sure what is going on here - do I need to install the moose-tools (mamba) to get the test harness work OK? Looks like things go wonky at the various __init__.py

**Comment by GiudGiud** [comment]
Hello

installing moose-tools will solve the first one. 
the undefined symbol is more problematic. It s likely the compiler suite was not consistent during the whole build process. Or that there are conflicts between several compilers available 

Guillaume

**Comment by mangerij** [comment]
@heinono1  what version of python is that build using? I have noticed that sometimes with older python versions the new moose can compile just fine (with little to no error) but then when you try to run the executables then you get hit errors.

**Reply by milljm** [reply]
If you are building from scratch (that is you are using system compilers), then you will need to install the python-dev package (or maybe it is called python3-dev). Hit requires Python Headers to build. Also, I would refrain from installing the moose-tools package, that package is tied to moose-mpich, which would cause a conflict between your OpenMPI and our package providing MPICH.

**Comment by heinono1** [comment]
Thanks for responses. I am using python 3-9.12 but I did not install the
dev package - I'll do that.

Cheers,
Olle

On Tue, Jun 7, 2022 at 7:16 AM Jason Miller ***@***.***>
wrote:

> If you are building from scratch (that is you are using system compilers),
> then you will need to install the python-dev package (or maybe it is called
> python3-dev). Hit requires Python Headers to build. Also, I would refrain
> from installing the moose-tools package, that package is tied to
> moose-mpich, which would cause a conflict between your OpenMPI and our
> package providing MPICH.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/discussions/21218#discussioncomment-2897584>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEKZEF5NSE2GGZUPJBKF35LVN44SFANCNFSM5YA4YKWA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
Olle Heinonen
***@***.***

**Comment by heinono1** [comment]
Installing python3-dev did not help. Note that I have deactivated conda env
to try to avoid conflicts with conda packages (I did not install
moose-tools). The problem now is that run_tests is looking for python in
/usr/bin/env and that directory is empty. Not sure how to make the test
harness point to the right executables.

On Tue, Jun 7, 2022 at 8:18 AM Olle Heinonen ***@***.***>
wrote:

> Thanks for responses. I am using python 3-9.12 but I did not install the
> dev package - I'll do that.
>
> Cheers,
> Olle
>
> On Tue, Jun 7, 2022 at 7:16 AM Jason Miller ***@***.***>
> wrote:
>
>> If you are building from scratch (that is you are using system
>> compilers), then you will need to install the python-dev package (or maybe
>> it is called python3-dev). Hit requires Python Headers to build. Also, I
>> would refrain from installing the moose-tools package, that package is tied
>> to moose-mpich, which would cause a conflict between your OpenMPI and our
>> package providing MPICH.
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/idaholab/moose/discussions/21218#discussioncomment-2897584>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AEKZEF5NSE2GGZUPJBKF35LVN44SFANCNFSM5YA4YKWA>
>> .
>> You are receiving this because you were mentioned.Message ID:
>> ***@***.***>
>>
>
>
> --
> Olle Heinonen
> ***@***.***
>


-- 
Olle Heinonen
***@***.***

**Reply by milljm** [reply]
`/usr/bin/env` does not exist? Can you find out where it is?

```bash
which env
```
whatever that returns, I'd like you to make a manual edit to the `moose/test/run_tests` script. And modify the shebang to match the path returned by the above command.

**Reply by milljm** [reply]
It could also be that your OS defaults to Python2? This has been something of debate in our group (and on the global stage) as to what a script _should_ do... to run the correct interpreter. My vote is:

```bash
#!/usr/bin/env python3
``` 

If you can make that additional change to `moose/test/run_test` and try again.?.

**Reply by heinono1** [reply]
Hi. /usr/bin/env does exist, and which env returns /usr/bin/env. The edit you suggested seems to work (the tests are running so far).

Thanks!

**Reply by heinono1** [reply]
THere are some crashes because /bin/sh cannot find python

**Reply by milljm** [reply]
I am betting there are more instances of the shebang in our scripts which is causing you grief. I have not tried this myself, but perhaps you can create your own symlink to python3 and call it python...

```bash
which python3
```
whatever that returns try this:
```bash
mkdir ~/bin; cd ~/bin
ln -s `which python3` python
export PATH=~/bin:$PATH
```
and re-run the tests? Basically we're tricking our scripts into using `python` which will be mapped to python3.

**Comment by heinono1** [comment]
Yeah I was thinking of doing that....

On Tue, Jun 7, 2022 at 9:03 AM Jason Miller ***@***.***>
wrote:

> I am betting there are more instances of the shebang in our scripts which
> is causing you grief. I have not tried this myself, but perhaps you can
> create you own symlink to python3 and call it python...
>
> which python3
>
> whatever that returns try this:
>
> mkdir ~/bin; cd ~/bin
> ln -s `which python3` pythonexport PATH=~/bin:$PATH
>
> and re-run the tests? Basically we're tricking our scripts into using
> python which will be mapped to python3.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/discussions/21218#discussioncomment-2898328>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEKZEFYWEEKRV6JOIGIQ37LVN5JDHANCNFSM5YA4YKWA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
Olle Heinonen
***@***.***

**Reply by milljm** [reply]
If that works, I think its time we change our shebangs...
@dschwen 
@permcody

**Comment by heinono1** [comment]
👍

On Tue, Jun 7, 2022 at 9:07 AM Jason Miller ***@***.***>
wrote:

> If that works, I think its time we change our shebangs...
> @dschwen <https://github.com/dschwen>
> @permcody <https://github.com/permcody>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/discussions/21218#discussioncomment-2898350>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEKZEF6TLJPTL2Y4VVSX3SDVN5JQ7ANCNFSM5YA4YKWA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
Olle Heinonen
***@***.***

**Comment by heinono1** [comment]
THere are some other issues as well (SKIP, not CRASH) - missing pandas,
deepdiff

On Tue, Jun 7, 2022 at 9:10 AM Olle Heinonen ***@***.***>
wrote:

> 👍
>
> On Tue, Jun 7, 2022 at 9:07 AM Jason Miller ***@***.***>
> wrote:
>
>> If that works, I think its time we change our shebangs...
>> @dschwen <https://github.com/dschwen>
>> @permcody <https://github.com/permcody>
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/idaholab/moose/discussions/21218#discussioncomment-2898350>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AEKZEF6TLJPTL2Y4VVSX3SDVN5JQ7ANCNFSM5YA4YKWA>
>> .
>> You are receiving this because you were mentioned.Message ID:
>> ***@***.***>
>>
>
>
> --
> Olle Heinonen
> ***@***.***
>


-- 
Olle Heinonen
***@***.***

**Reply by GiudGiud** [reply]
install those libraries and it will stop skipping

**Comment by heinono1** [comment]
The symlink to python3 causes crashes - too many levels of symbolic links

On Tue, Jun 7, 2022 at 9:03 AM Jason Miller ***@***.***>
wrote:

> I am betting there are more instances of the shebang in our scripts which
> is causing you grief. I have not tried this myself, but perhaps you can
> create you own symlink to python3 and call it python...
>
> which python3
>
> whatever that returns try this:
>
> mkdir ~/bin; cd ~/bin
> ln -s `which python3` pythonexport PATH=~/bin:$PATH
>
> and re-run the tests? Basically we're tricking our scripts into using
> python which will be mapped to python3.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/discussions/21218#discussioncomment-2898328>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEKZEFYWEEKRV6JOIGIQ37LVN5JDHANCNFSM5YA4YKWA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
Olle Heinonen
***@***.***

**Comment by heinono1** [comment, marked as answer]
I think making symlink directly to /usr/bin/python3.10 may have worked

On Tue, Jun 7, 2022 at 9:15 AM Olle Heinonen ***@***.***>
wrote:

> The symlink to python3 causes crashes - too many levels of symbolic links
>
> On Tue, Jun 7, 2022 at 9:03 AM Jason Miller ***@***.***>
> wrote:
>
>> I am betting there are more instances of the shebang in our scripts which
>> is causing you grief. I have not tried this myself, but perhaps you can
>> create you own symlink to python3 and call it python...
>>
>> which python3
>>
>> whatever that returns try this:
>>
>> mkdir ~/bin; cd ~/bin
>> ln -s `which python3` pythonexport PATH=~/bin:$PATH
>>
>> and re-run the tests? Basically we're tricking our scripts into using
>> python which will be mapped to python3.
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/idaholab/moose/discussions/21218#discussioncomment-2898328>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AEKZEFYWEEKRV6JOIGIQ37LVN5JDHANCNFSM5YA4YKWA>
>> .
>> You are receiving this because you were mentioned.Message ID:
>> ***@***.***>
>>
>
>
> --
> Olle Heinonen
> ***@***.***
>


-- 
Olle Heinonen
***@***.***

**Reply by milljm** [reply]
Good to know! I have created an issue I plan on implementing soon: https://github.com/idaholab/moose/issues/21235

**Comment by heinono1** [comment]
Hi Guillaume,

Thanks. I do not think there are inconsistencies in the compiler suite or
that there are competing compilers - I had just installed compilers and
wrappers, and I also made sure to set environment variables (FC, CPP, CC
etc) to point unambiguously to the compilers and wrappers. I'll fiddle with
the moose-tools to see if I can get at least most of them.

Cheers,
Olle

On Mon, Jun 6, 2022 at 6:17 PM Guillaume Giudicelli <
***@***.***> wrote:

> Hello
>
> installing moose-tools will solve the first one.
> the undefined symbol is more problematic. It s likely the compiler suite
> was not consistent during the whole build process. Or that there are
> conflicts between several compilers available
>
> Guillaume
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/discussions/21218#discussioncomment-2893932>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEKZEFYIHBRHS6BBMKZH2GLVN2BHXANCNFSM5YA4YKWA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
Olle Heinonen
***@***.***

**Reply by milljm** [reply]
Looks like I let this shebang PR slip my mind. I'll start on that right now...
